Diploma API

piskov opened the pull request:
Попытался сделать все требования (включая чеклист на пятерку).

piskov committed:
API logic

piskov committed:
Return _id in article DTO

Boortcore commented:
# Резюме:
## На тройку
- [x] В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id.
- [x] Нет ошибок линтера.
- [x] В разделе scripts файла package.json:
  - [x] есть команда npm run start, которая запускает сервер на localhost:3000;
  - [x] и npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] Когда все зависимости установлены, приложение запускается командой npm run dev без ошибок.
- [x] Описанные роуты работают корректно:
  - [x] запрос на GET /users/me возвращает информацию о пользователе (email и имя);
  - [x] GET /articles — все сохранённые статьи;
  - [ ] POST /articles создаёт статью с переданными в теле данными;
![image](https://user-images.githubusercontent.com/17438227/74179848-4f03d500-4c4f-11ea-9ef3-8032868e20f3.png)
![image](https://user-images.githubusercontent.com/17438227/74179879-617e0e80-4c4f-11ea-932f-95ba8261279c.png)

  - [x] DELETE /articles/articleId удаляет статью по _id; 
  - [x] POST /signup создаёт пользователя с переданными в теле данными;
  - [x] POST /signin возвращает JWT, если в теле запроса переданы правильные почта и пароль.
- [x] Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] Роуты пользователей и роуты статей описаны в отдельных файлах.
- [ ] Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
маршруты должны быть описаны в файле `index.js` папки `routes`, но не в корневом файле `app.js`
![image](https://user-images.githubusercontent.com/17438227/74180081-d8b3a280-4c4f-11ea-90f3-03e28c50c33a.png)

- [x] Ошибки API обрабатываются:
  - [x] если в запросе что-то не так, сервер возвращает ответ с ошибкой и её статусом;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] В production-режиме адрес базы данных берётся из process.env.
- [x] Реализовано бережное хранение пароля:
  - [x] пароль хранится в зашифрованном виде;
  - [x] API не возвращает хеш пароля клиенту.
- [x] Данные валидируются перед добавлением в базу.
- [x] Пользователь не может удалить сохранённую карточку из профиля другого пользователя.
- [x] К серверу можно обратиться и по http, и по https, обращаясь к домену, указанному в README.md.
- [x] Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.
  
## На четвёрку
- [x] Асинхронные операции реализованы промисами или async/await.
- [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры, проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
- [x] Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] Ошибки обрабатываются централизованным обработчиком.
- [ ] Централизованный обработчик ошибок описан в отдельном модуле.
- [ ] В случае ошибки API возвращает статус ответа, который соответствует типу ошибки.
Статус ошибки должен быть `403`
![image](https://user-images.githubusercontent.com/17438227/74179994-a2762300-4c4f-11ea-8055-392c4f588fec.png)
- [x] К серверу можно обратиться по https по адресу домена, указанному в README.md.
- [x] API приложения располагается на домене вида: name.zone/api, а не просто name.zone:

## На пятёрку
- [x] Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [x] Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [x] секретный ключ для JWT в режиме разработки вынесен в отдельный конфигурационный файл;
  - [x] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [x] Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] API размещено на отдельном поддомене, например: api.news-explorer.tk.

piskov committed:
Code refactoring

- signin and login routes in separate files
- 404 and centralized error handling in separate files
- add Forbidden error support

Boortcore commented:
# Резюме:
## На тройку
- [x] В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id.
- [x] Нет ошибок линтера.
- [x] В разделе scripts файла package.json:
  - [x] есть команда npm run start, которая запускает сервер на localhost:3000;
  - [x] и npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] Когда все зависимости установлены, приложение запускается командой npm run dev без ошибок.
- [x] Описанные роуты работают корректно:
  - [x] запрос на GET /users/me возвращает информацию о пользователе (email и имя);
  - [x] GET /articles — все сохранённые статьи;
  - [x] POST /articles создаёт статью с переданными в теле данными;
  - [x] DELETE /articles/articleId удаляет статью по _id; 
  - [x] POST /signup создаёт пользователя с переданными в теле данными;
  - [x] POST /signin возвращает JWT, если в теле запроса переданы правильные почта и пароль.
- [x] Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] Роуты пользователей и роуты статей описаны в отдельных файлах.
- [x] Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] Ошибки API обрабатываются:
  - [x] если в запросе что-то не так, сервер возвращает ответ с ошибкой и её статусом;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] В production-режиме адрес базы данных берётся из process.env.
- [x] Реализовано бережное хранение пароля:
  - [x] пароль хранится в зашифрованном виде;
  - [x] API не возвращает хеш пароля клиенту.
- [x] Данные валидируются перед добавлением в базу.
- [x] Пользователь не может удалить сохранённую карточку из профиля другого пользователя.
- [x] К серверу можно обратиться и по http, и по https, обращаясь к домену, указанному в README.md.
- [x] Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.
  
## На четвёрку
- [x] Асинхронные операции реализованы промисами или async/await.
- [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры, проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
- [x] Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] Ошибки обрабатываются централизованным обработчиком.
- [x] Централизованный обработчик ошибок описан в отдельном модуле.
- [x] В случае ошибки API возвращает статус ответа, который соответствует типу ошибки.
- [x] К серверу можно обратиться по https по адресу домена, указанному в README.md.
- [x] API приложения располагается на домене вида: name.zone/api, а не просто name.zone:

## На пятёрку
- [x] Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [x] Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [x] секретный ключ для JWT в режиме разработки вынесен в отдельный конфигурационный файл;
  - [x] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [x] Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] API размещено на отдельном поддомене, например: api.news-explorer.tk.
